File: Week10/rst.cpp

#include <bits/stdc++.h>
using namespace std;
/*
  Brian Grenier
  1545276
  bgrenier

  List any resources you used below (eg. urls, name of the algorithm from our code archive).
  Remember, you are permitted to get help with general concepts about algorithms
  and problem solving, but you are not permitted to hunt down solutions to
  these particular problems!

    https://eclass.srv.ualberta.ca/pluginfile.php/7035320/mod_resource/content/3/tourists.cpp

  List any classmate you discussed the problem with. Remember, you can only
  have high-level verbal discussions. No code should be shared, developed,
  or even looked at in these chats. No formulas or pseudocode should be
  written or shared in these chats.

  N/A

  By submitting this code, you are agreeing that you have solved in accordance
  with the collaboration policy in CMPUT 403.
*/

int n, Q, u, v;

typedef long long ll;
#define MAXN 200010
#define LOGN 18
// the graph
vector<int> g[MAXN];
int up[LOGN][MAXN], depth[MAXN];

// Find the vertex that is 2^i steps above u
void dfs(int v, int pre, int d) {
    depth[v] = d;

    for (auto w : g[v])
        if (w != pre)
            dfs(w, v, d+1);

    up[0][v] = pre;
}

// compute the LCA of a and b
int lca(int a, int b) {
    if (depth[a] > depth[b]) swap(a, b);

    // raise b to have the same depth as a by jumping up via
    // powers of 2 (sort of like the fast exponentiation trick)
    for (int j = LOGN-1; j >= 0; --j)
        if (depth[a] + (1<<j) <= depth[b])
            b = up[j][b];

    // then a was an ancestor of b, so it is the LCA
    if (a == b) return a;

    // otherwise, simultaneously raise them both up via the powers of
    // 2 trick but only if this does not cause them to be equal
    for (int j = LOGN-1; j >= 0; --j)
        if (up[j][a] != up[j][b]) {
            a = up[j][a];
            b = up[j][b];
        }

    // now a != b but they have a common parent, which is the LCA
    return up[0][a];
}

// just the forumula from the file header
ll dist(ll a, ll b) {
    return depth[a] + depth[b] - 2*depth[lca(a, b)] + 1;
}



int main() {
    cin >> n >> Q;

    for (int i = 0; i < n-1; ++i) {
        cin >> u >> v;
        --u;
        --v;
        g[u].push_back(v);
        g[v].push_back(u);
    }

    for (auto x : up) fill(x, x+n, -1);
    dfs(0, -1, 0);

     for (int j = 1; j < LOGN; ++j)
        for (int v = 0; v < n; ++v)
            if (up[j-1][v] != -1)
                up[j][v] = up[j-1][up[j-1][v]];

    for (int i = 0; i < Q; ++i) {
        cin >> u >> v;
        --u;
        --v;
        ll z = dist(u,v);
        cout << ((z*(z-1))/2)+n << endl;
        
    }
}

File: Week10/rst.py
"""
  Brian Grenier
  1545276
  bgrenier

  List any resources you used below (eg. urls, name of the algorithm from our code archive).
  Remember, you are permitted to get help with general concepts about algorithms
  and problem solving, but you are not permitted to hunt down solutions to
  these particular problems!

    https://github.com/UAPSPC/Code-Archive/blob/master/graph/lca.cpp

  List any classmate you discussed the problem with. Remember, you can only
  have high-level verbal discussions. No code should be shared, developed,
  or even looked at in these chats. No formulas or pseudocode should be
  written or shared in these chats.

  N/A

  By submitting this code, you are agreeing that you have solved in accordance
  with the collaboration policy in CMPUT 403.
"""

import sys
from functools import lru_cache
from math import log2

class lca:
    def __init__(self, n):
        self.P = [[-1] * (self.log2(n)+1)] * (n+1)
        self.D = [-1] * (n+1)
        self.P[0][0] = -2
        self.n = n

    # @lru_cache(maxsize=None)
    def log2(self, n):
        return int(log2(n))

    # @lru_cache(maxsize=None)
    def parent(self, u, i):
        v = self.P[u][i]
        if v != -1: return v
        if i <= 0: return self.P[u][0]
        if self.parent(u, i-1) == -2:
            self.P[u][i] = -2
            return -2
        self.P[u][i] = self.parent(self.parent(u, i-1), i-1)
        return self.P[u][i]

    # @lru_cache(maxsize=None)
    def depth(self, u):
        if self.D[u] != -1: return self.D[u]
        if u > 0:
            self.D[u] = 1+self.depth(self.P[u][0])
            return self.D[u]
        else:
            return 0

    # @lru_cache(maxsize=None)
    def lca(self, u, v):
        if self.depth(v) > self.depth(u): return self.lca(v, u)
        for i in range(self.log2(n), -1, -1):
            if (self.depth(u) - (1 << i)) >= self.depth(v): u = self.parent(u,i)
        if u == v: return u
        for i in range(self.log2(n)-1, -1, -1):
            if self.parent(u,i) != self.parent(v,i):
                u = self.parent(u,i)
                v = self.parent(v,i)

        return self.parent(u,0)

    # @lru_cache(maxsize=None)
    def dist(self, u,v):
        a = self.lca(u,v)
        return self.depth(u) + self.depth(v) - 2*self.depth(a)+1

if __name__ == '__main__':
    sys.setrecursionlimit(1500)
    n, q = map(int, sys.stdin.readline().split())
    l = lca(n)

    for i in range(n-1):
        u, v = map(int, sys.stdin.readline().split())
        l.P[v][0] = u

    for i in range(q):
        u, v = map(int, sys.stdin.readline().split())

        d = l.dist(u,v)
        print((d-1) * max(u,v)+1)